T18983

smspillaz opened the pull request:
https://phabricator.endlessm.com/T18983

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -291,24 +296,32 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
-  if (!id_is_valid)
+  /* The assumption is that this will not be called
+   * multiple times for the same instance across different
+   * threads. */
```
Is this assumption true? Was the lock not there for a reason before?

I believe the answers are “yes; no”; and so I don't think this comment is necessary. There is plenty of non-thread-safe code in this application.

smspillaz replied:
Yeah, the lock being there did seem strange (what was more strange was the fact that we were using a static var for this). In any case, I'm almost certain that you're right - just put that comment there in case anyone would pick up on it ;-)

wjt commented on data/eos-metrics-event-recorder.service.in:
```diff
@@ -9,6 +9,7 @@ User=metrics
 Type=dbus
 BusName=com.endlessm.Metrics
 ExecStart=@libexecdir@/eos-metrics-event-recorder
+ExecReload=/bin/kill -SIGHUP $MAINPID
```
`man systemd.service` actually recommends against this approach – it suggests using a mechanism which allows you to wait for it to complete (and see if it was successful). An example from `dbus.service` itself:

```
/lib/systemd/system/dbus.service:ExecReload=/usr/bin/dbus-send --print-reply --system --type=method_call --dest=org.freedesktop.DBus / org.freedesktop.DBus.ReloadConfig
```

(I suggested the `SIGHUP` approach for `gdm` because it already implemented a `SIGHUP` handler.)

But I appreciate that this approach does not involve adding methods to the API (defined in a separate package) etc etc.

smspillaz replied:
I suppose we could add some new d-bus API to do this, it wouldn't be too hard.

At that point I might as well add some API to wipe the existing metrics cache too.

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -291,24 +296,32 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+
+  g_message ("EmerMachineIdProvider: Reloading from %s", priv->path);
```
This message isn't true – that file will not be read until an unspecified time in the future.

smspillaz replied:
That's true. I suppose I could rephrase it to "will reload from %s"

wjt commented on daemon/emer-daemon.c:
```diff
@@ -1547,6 +1547,15 @@ emer_daemon_new (const gchar             *persistent_cache_directory,
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+
+  if (priv->machine_id_provider)
```
I prefer `if (priv->machine_id_provider != NULL)` for checks like this, and a quick skim of this file suggests that's the style in use here.

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -165,10 +166,14 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
 }
 
-// Mandatory empty function.
+/* Explicitly set id_is_valid to false */
```
I don't think this comment is needed.

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -165,10 +166,14 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+
+  priv->id_is_valid = FALSE;
```
I don't really think this is needed – glib guarantees that the priv structure will be 0ed, and the other two fields don't get explicitly initialised here – but ok.

smspillaz replied:
Strictly speaking, that's true, though my preference is to be explicit about these kinds of things.

smspillaz commented:
@wjt I addressed most of your comments - thanks for the quick feedback :)

wjt commented on daemon/emer-main.c:
```diff
@@ -200,6 +209,13 @@ quit_main_loop (GMainLoop *main_loop)
 }
 
+static gboolean
+reload_daemon_config (EmerDaemon *daemon)
```
I think this function is now unused.

wjt commented on daemon/emer-main.c:
```diff
@@ -222,6 +238,8 @@ on_bus_acquired (GDBusConnection *system_bus,
   g_signal_connect (server, "handle-upload-events",
                     G_CALLBACK (on_upload_events), daemon);
+  g_signal_connect (server, "handle-reload-config",
+                    G_CALLBACK (on_reload_config), daemon);
```
Should this be a privileged operation, ie. added to `on_authorize_method_check`? You could make an argument that calling this method has no observable effect if `/etc/machine-id` hasn't changed, so it does not need to be authorized.

smspillaz commented:
Okay, reworked the approach here considering that we can't just nuke the machine-id.

I added a `ResetTrackingId` method to the daemon which calls through a system helper to write an override file in `/etc/eos-event-recorder-daemon/machine-id-override`. If that file is present, its value is preferred to that of `/etc/machine-id`. This will allow the demo mode to reset the machine-id safely.

wjt commented on system-helper/eos-event-recorder-system-helper.c:
```diff
@@ -0,0 +1,195 @@
+
+#include "emer-metrics-system-helper.h"
+
+#define TRACKING_ID_OVERRIDE SYSCONFDIR "/eos-metrics-event-recorder/machine-id-override"
```
A better choice of path would be `/etc/metrics/machine-id-override`. The daemon already writes a file to this directory ([`cache-size.conf`](https://github.com/endlessm/eos-event-recorder-daemon/blob/master/daemon/emer-cache-size-provider.c#L39); `eos-metrics-permissions.conf` is also in that directory but for reasons that elude me it is edited by external scripts and re-read by the daemon every time its contents are needed). The `metrics` user owns this directory so there is no need for a privileged helper.

smspillaz replied:
Ah of course! That would allow me to drop the system helper too...

wjt commented on Makefile.am:
```diff
@@ -55,6 +55,33 @@ noinst_PROGRAMS =
+
+# systemd service units
+#
+# Defined as systemdsystemunitdir in ../configure.ac.
```
The relative path in this comment is now wrong since this file is in the same directory as `configure.ac`.

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -78,23 +96,23 @@ static GParamSpec *emer_machine_id_provider_props[NPROPS] = { NULL, };
-set_id_path (EmerMachineIdProvider *self,
-             const gchar           *given_path)
+set_id_paths (EmerMachineIdProvider *self,
+              const GStrv            given_paths)
```
If you make this take ownership a non-const `GStrv`…

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -224,62 +250,118 @@ hyphenate_uuid (gchar *uuid_sans_hyphens)
+    }
+
+  uuid_copy (priv->id, id);
+  uuid_clear (id);
```
This `uuid_clear` is not needed, it's just a stack-allocated array.

wjt replied:
Still true!

wjt commented on tests/daemon/test-machine-id-provider.c:
```diff
@@ -31,7 +31,9 @@
 #define HYPHENS_IN_ID 4
 
 #define TESTING_FILE_PATH "/tmp/testing-machine-id"
+#define TESTING_OVERRIDE_FILE_PATH "/tmp/override-testing-machine-id"
```
I appreciate the existing test does the same but I don't really like using fixed paths in `/tmp` here. I think the fixture setup should create the temporary files.

wjt commented on tests/daemon/test-machine-id-provider.c:
```diff
@@ -31,7 +31,9 @@
 #define TESTING_FILE_PATH "/tmp/testing-machine-id"
+#define TESTING_OVERRIDE_FILE_PATH "/tmp/override-testing-machine-id"
 #define TESTING_ID        "04448f74fde24bd7a16f8da17869d5c3\n"
+#define TESTING_OVERRIDE_ID        "04448f74fde24bd7a16f8da17869d5c4\n"
```
Might be nice to generate a more-distinct UUID so that assertion failures are clearer, rather than one which differs in only 2 trailing bits!

wjt commented on tests/daemon/test-machine-id-provider.c:
```diff
@@ -102,8 +106,55 @@ static void
+  g_assert (emer_machine_id_provider_get_id (id_provider, id));
+  gchar unparsed_id[HYPHENS_IN_ID + FILE_LENGTH];
+  uuid_unparse_lower (id, unparsed_id);
+  gchar* unhypenated_id = unhyphenate_uuid (unparsed_id);
```
g_autofree here and remove g_free below

wjt commented on tests/daemon/test-machine-id-provider.c:
```diff
@@ -102,8 +106,55 @@ static void
+  gchar* unhypenated_id = unhyphenate_uuid (unparsed_id);
+  g_assert_cmpstr (TESTING_ID, ==, unhypenated_id);
+  g_free (unhypenated_id);
+  g_object_unref (id_provider);
```
If you were feeling keen you could add `G_DEFINE_AUTOPTR_CLEANUP_FUNC (EmerMachineIdProvider, g_object_unref);` to `emer-machine-id-provider.h` and then use `g_autoptr(EmerMachineIdProvide) id_provider` above.

wjt commented on tests/daemon/test-machine-id-provider.c:
```diff
@@ -128,6 +179,10 @@ main (gint                argc,
+  ADD_CACHE_TEST_FUNC ("/machine-id-provider/can-get-id-override",
+                       test_machine_id_provider_can_get_id_override);
+  ADD_CACHE_TEST_FUNC ("/machine-id-provider/can-get-id-override-not-exist",
+                       test_machine_id_provider_can_get_id_override_not_exist);
```
Please also test the case where the first file exists but is malformed.

wjt commented on Makefile.am:
```diff
@@ -62,11 +62,6 @@ substitute_libexecdir = sed \
-# systemd service units
-#
-# Defined as systemdsystemunitdir in ../configure.ac.
-dist_systemdsystemunit_DATA = 
```
This  was only just moved to `Makefile.am` two patches ago!

smspillaz replied:
`git rebase` 😿😿😿😿

wjt commented on Makefile.am:
```diff
@@ -77,7 +72,7 @@ dist_systembussecuritypolicy_DATA =
 
 # policykit
 policydir = ${datadir}/polkit-1/actions
-dist_policy_DATA = data/com.endlessm.Metrics.policy
```
Should this not be part of the previous patch?

wjt commented on Makefile.am:
```diff
@@ -86,6 +81,9 @@ persistentcachedir=$(localstatedir)/cache/metrics/
 permissions_file=$(configdir)eos-metrics-permissions.conf
 
+# systemd
+systemdsystemunit_DATA = 
```
What's this doing here?

smspillaz replied:
We append to it later in one of the subdirectories included by this makefile

wjt commented on tests/daemon/test-daemon.c:
```diff
@@ -77,6 +78,8 @@ typedef struct _Fixture
   gint64 absolute_time;
   gchar *request_path;
+
+  gchar *tempfiles_dir;
```
This appears to be unused.

wjt commented on tests/daemon/test-daemon.c:
```diff
@@ -924,13 +927,17 @@ create_test_object (Fixture *fixture)
        gconstpointer unused)
 {
+  g_autoptr(GFileIOStream) stream = NULL;
+  g_autoptr(GError) error = NULL;
```
These variables are not used.

wjt commented on daemon/emer-daemon.c:
```diff
@@ -1547,6 +1580,29 @@ emer_daemon_new (const gchar             *persistent_cache_directory,
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+
+  if (priv->override_machine_id_path != NULL)
```
Given the point of `EmerMachineIdProvider` is to abstract away how the machine ID is stored, I think it should have an `emer_machine_id_provider_reset_tracking_id` function. It already knows what the override path is, and the call to `emer_machine_id_provider_reload` to tell it the world has changed underneath its feet suggests that it should take care of the reset, too. This would also make the test case simpler because you don't need to hit the disk at all, and you can just assert that the mock version of this function was called.

smspillaz replied:
This is certainly a tension I was feeling too when implementing this. `EmerMachineIdProvider` takes a strv of potential override paths in priority order, so it can't really know which path to write to.

That is unless what we want to do is only have a single override path, which could be a smart move.

wjt commented on daemon/emer-daemon.c:
```diff
@@ -1547,6 +1580,29 @@ emer_daemon_new (const gchar             *persistent_cache_directory,
+  if (priv->machine_id_provider != NULL)
+    emer_machine_id_provider_reload (priv->machine_id_provider);
+
+  return TRUE;
```
If there were a signal that `eos-metrics-instrumentation` could use to discard all its state and re-emit the boot-time events, then I think the `ResetTrackingId` method should discard all buffered events and then emit that signal. But it's both or neither: if we discard all cached events here, we will receive the "it's demo mode!" event (which is emitted from g-i-s only after this method returns) but not the boot-time events with info like the OS version, image, hardware, etc.

wjt replied:
(I'm not proposing adding that signal and behaviour here – I think that's a separate task.)

smspillaz commented on daemon/emer-machine-id-provider.c:
```diff
@@ -60,12 +64,26 @@ G_DEFINE_TYPE_WITH_PRIVATE (EmerMachineIdProvider, emer_machine_id_provider, G_T
 static GParamSpec *emer_machine_id_provider_props[NPROPS] = { NULL, };
 
+static GStrv
+dup_strv (GStrv input)
```
Ah damn, I was looking for that and must have missed it. Thanks!

wjt commented on tests/daemon/test-machine-id-provider.c:
```diff
@@ -72,25 +80,41 @@ unhyphenate_uuid (gchar *uuid_with_hyphens)
-test_machine_id_provider_new_succeeds (gboolean     *unused,
-                                       gconstpointer dontuseme)
+test_machine_id_provider_new_succeeds (MachineIdTestFixture *fixture,
+                                       gconstpointer          dontuseme)
```
Mismatched alignment

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -156,24 +169,35 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
-    g_param_spec_string ("path", "Path",
-                         "The path to the file where the unique identifier is stored.",
+    g_param_spec_string ("path", "Machine ID path",
+                         "The path to where the immutable machine id is stored.",
```
`id` here should be capitalised

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -156,24 +169,35 @@ emer_machine_id_provider_class_init (EmerMachineIdProviderClass *klass)
                          G_PARAM_WRITABLE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+  emer_machine_id_provider_props[PROP_OVERRIDE_PATH] =
+    g_param_spec_string ("override-path", "Tracking ID override path",
+                         "The path to where a mutable tracking id override path is stored.",
```
Here too

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -224,62 +254,119 @@ hyphenate_uuid (gchar *uuid_sans_hyphens)
+      if (!read_one_machine_id (priv->path, id, &local_error))
+        {
+          g_message ("Failed to read machine id %s: %s",
+                      priv->override_path,
```
This is the wrong path.

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -224,62 +250,118 @@ hyphenate_uuid (gchar *uuid_sans_hyphens)
+
+  if (uuid_is_null (id))
+    {
+      g_critical ("Failed to read in a unique machine id");
```
I still think a better message would be something like:

> Failed to read machine ID from either %s or %s

for the two paths (override, then fallback)

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -291,24 +378,64 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
+          return FALSE;
+        }
+
+      g_clear_error (&local_error);
```
Unneeded since `local_error` is not re-used

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -291,24 +378,64 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
+
+  g_message ("EmerMachineIdProvider: Will reload from:\n%s", priv->override_path);
+
+  priv->id_is_valid = FALSE;
```
This function needs to `return TRUE` on the happy path. I'd hope the compiler would issue a warning.

smspillaz replied:
It does not, unfortunately, I have noticed :(

wjt commented on daemon/emer-machine-id-provider.c:
```diff
@@ -291,24 +378,64 @@ emer_machine_id_provider_get_id (EmerMachineIdProvider *self,
+  if (!write_tracking_id_file (priv->override_path, error))
+    return FALSE;
+
+  g_message ("EmerMachineIdProvider: Will reload from:\n%s", priv->override_path);
```
I think the newline should be removed

wjt commented on data/com.endlessm.Metrics.policy:
```diff
@@ -38,4 +38,14 @@
     </defaults>
   </action>
 
+  <action id="com.endlessm.Metrics.ResetTrackingId">
```
Is this used? I don't see any reference to it. I think you need to update `on_authorize_method_check`.

smspillaz replied:
Ah! looks like I overzealously reverted this change. I'll unrevert it, since we should make resetting the tracking ID a privileged operation.

wjt commented on data/com.endlessm.Metrics.policy:
```diff
@@ -38,4 +38,14 @@
   </action>
 
+  <action id="com.endlessm.Metrics.ResetTrackingId">
+    <description>Reset the tracking id used by the metrics system</description>
```
ID should be capitalized

wjt commented:
You'll also need to rebase the branch because it now conflicts with master.

smspillaz commented:
Addressed all comments. Also looks like a file was missing on master, so I added that.

Should be able to squash down all the commits into 9eeba0d once we're done, just included them here for completeness.

wjt commented on daemon/emer-main.c:
```diff
@@ -134,6 +151,58 @@ on_upload_events (EmerEventRecorderServer *server,
+static const gsize n_not_authorized_error_messages = G_N_ELEMENTS (not_authorized_error_messages);
+
+static const gchar *
+not_authorized_error_message_for_method (AuthorizedMethod method)
```
This method is never used!

I think a better approach here would be to define:

```c
typedef struct {
    const gchar *method_name;
    const gchar *method_full_name;
    const gchar *error_message;
} AuthorizedMethod;

static const AuthorizedMethods[] authorized_methods = {
   { "SetEnabled", "com.endlessm.Metrics.SetEnabled", "..." },
   { "ResetTrackingId", ... },
   { NULL }
};

static const AuthorizedMethod *
lookup_authorized_method (const gchar *method_name)
{
  // obvious loop, return NULL if not found
}
```

Then `on_authorize_method_check` would call this function once, early-return if it returned NULL, and then have all the other information it needed right there. No need for extra lookups and `g_assert`s.
```

smspillaz replied:
Makes sense, I think that's a lot simpler.

smspillaz commented:
Okay, rebased on #225, squashed all the commits except the last one with the last review comment together.

wjt commented on daemon/emer-main.c:
```diff
@@ -186,8 +239,8 @@ on_authorize_method_check (GDBusInterfaceSkeleton *interface,
-                                           "Disabling metrics is only "
-                                           "allowed from system settings");
+                                           "%s",
+                                           authorized_method->error_message);
```
If you were feeling really keen you could use `g_dbus_method_invocation_return_error_literal` and avoid the `"%s"`.

wjt commented on Makefile.am:
```diff
@@ -98,6 +126,9 @@ EOS_COVERAGE_BLACKLIST_PATTERNS =	\
 AM_LDFLAGS += @EOS_C_COVERAGE_LDFLAGS@
 
+# # # COMMON LIBRARY # # #
+include $(top_srcdir)/shared/Makefile.am.inc
```
```
automake: error: cannot open < shared/Makefile.am.inc: No such file or directory
autoreconf: automake failed with exit status: 1
```

smspillaz replied:
Ouch, thanks for spotting - I rebased and put it into the right commit.

wjt commented on tests/daemon/mock-machine-id-provider.c:
```diff
@@ -48,16 +55,29 @@ emer_machine_id_provider_new (void)
+{
+  EmerMachineIdProviderPrivate *priv =
+    emer_machine_id_provider_get_instance_private (self);
+  priv->has_override = TRUE;
```
```
../../tests/daemon/mock-machine-id-provider.c: In function 'emer_machine_id_provider_reset_tracking_id':
../../tests/daemon/mock-machine-id-provider.c:83:1: error: control reaches end of non-void function [-Werror=return-type]
```

Will fix this…
